### PR TITLE
add Affinex MemRefDatFlowOpt pass

### PR DESCRIPTION
**Pull request by adstraw**

### CODE REVIEW

File: pmlc/dialect/affinex/tests/memref_dataflow_opt.mlir [hunk 1]
```diff
@@ -0,0 +1,181 @@
+// RUN: pmlc-opt -affinex-memref-dataflow-opt %s | FileCheck %s
+
+// CHECK: simple
+func @simple(%arg: memref<1xf32>) {
+  %cst = constant 0.000000e+00 : f32
+  // CHECK: affine.store
+  affine.store %cst, %arg[0] : memref<1xf32>
+  // CHECK-NOT: affine.load
+  %0 = affine.load %arg[0] : memref<1xf32>
+  // CHECK: addf %cst, %cst
```
Comment (flaub) [inline pmlc/dialect/affinex/tests/memref_dataflow_opt.mlir:10]: Generally, we don't want to use raw `%xyz` in CHECK statements. 
https://mlir.llvm.org/getting_started/TestingGuide/#filecheck-best-practices

File: pmlc/dialect/affinex/transforms/memref_dataflow_opt.cc [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    MemAccess access;
+    access.memref = DenseMapInfo<mlir::Value>::getEmptyKey();
+    return access;
+  }
+  static inline MemAccess getTombstoneKey() {
+    MemAccess access;
+    access.memref = DenseMapInfo<mlir::Value>::getEmptyKey();
+    return access;
+  }
+  static unsigned getHashValue(const MemAccess &Val) {
```
Comment (flaub) [inline pmlc/dialect/affinex/transforms/memref_dataflow_opt.cc:28]: rename `Val` to `val` (the older style from LLVM is deprecated).

File: pmlc/dialect/affinex/transforms/memref_dataflow_opt.cc [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "pmlc/dialect/affinex/transforms/pass_detail.h"
+
+using namespace mlir; // NOLINT
+
+struct MemAccess {
+  Value memref;
+  AffineMap map;
+  llvm::SmallVector<Value, 4> indices;
+  MemAccess() = default;
```
Comment (flaub) [inline pmlc/dialect/affinex/transforms/memref_dataflow_opt.cc:12]: Is this needed?

File: pmlc/dialect/affinex/transforms/passes.td [hunk 4]
```diff
@@ -0,0 +1,11 @@
+#ifndef __PMLC_DIALECT_AFFINEX_PASSES__
+#define __PMLC_DIALECT_AFFINEX_PASSES__
+
+include "mlir/Pass/PassBase.td"
+
+def AffinexMemRefDataFlowOpt : FunctionPass<"affinex-memref-dataflow-opt"> {
+  let summary = "Performs store to load forwarding to eliminate memory accesses and potentially the entire memref if all its accesses are forwarded.";
```
Comment (flaub) [inline pmlc/dialect/affinex/transforms/passes.td:7]: Limit `.td` files to 80 columns.

**Review by flaub — COMMENTED** [review]
Minor comments, but logic LGTM

File: pmlc/dialect/affinex/transforms/memref_dataflow_opt.cc (hunk 2 above)
Comment (adstraw) [reply pmlc/dialect/affinex/transforms/memref_dataflow_opt.cc:28]: Done

File: pmlc/dialect/affinex/transforms/memref_dataflow_opt.cc (hunk 3 above)
Comment (adstraw) [reply pmlc/dialect/affinex/transforms/memref_dataflow_opt.cc:12]: No, removed.

File: pmlc/dialect/affinex/transforms/passes.td (hunk 4 above)
Comment (adstraw) [reply pmlc/dialect/affinex/transforms/passes.td:7]: Done.

File: pmlc/dialect/affinex/tests/memref_dataflow_opt.mlir (hunk 1 above)
Comment (adstraw) [reply pmlc/dialect/affinex/tests/memref_dataflow_opt.mlir:10]: Done.

**Review by flaub — APPROVED** [review]

**Pull request merged**